pwndbg/commands/__init__.py

```diff
@@ -194,13 +258,93 @@ def __call__(self, *args: Any, **kwargs: Any) -> str | None:
+
+    Always use this to decorate your commands.
+    """
+
+    def __init__(
+        self,
+        parser_or_desc: argparse.ArgumentParser | str,
+        *,  # All further parameters are not positional
+        category: CommandCategory,
+        command_name: str | None = None,
+        aliases: List[str] = [],
+        only_debuggers: Set[pwndbg.dbg_mod.DebuggerType] = None,
+        exclude_debuggers: Set[pwndbg.dbg_mod.DebuggerType] = None,
+    ) -> None:
+        # Setup an ArgumentParser even if we were only passed a description.
+        if isinstance(parser_or_desc, str):
+            self.parser = argparse.ArgumentParser(description=parser_or_desc)
+        else:
+            assert isinstance(parser_or_desc, argparse.ArgumentParser)
+            self.parser = parser_or_desc
+
+        self.category = category
+        self.command_name = command_name
+        self.aliases = aliases
+        self.only_debuggers = only_debuggers
+        self.exclude_debuggers = exclude_debuggers
+
+    def __call__(self, function: Callable[..., Any]) -> CommandObj:
+        # Since this is the __call__ of a parametrized decorator, it is
+        # invoked during decoration, and it must return a callable object
+        # i.e. the "real" decorator of the function.
+
+        # If this command is not valid for this debugger, do not even
+        # pass it to ComandObj to be registered with the debugger API.
+        # Also make sure it raises an error if it is called from the code.
+        if self.only_debuggers is not None and pwndbg.dbg.name() not in self.only_debuggers:
+
+            def decorator(*args, **kwargs):
+                raise InvalidDebuggerError(
+                    f"This command cannot be used in {pwndbg.dbg.name()}.\n"
+                    f"It is only valid for {self.only_debuggers}."
+                )
+
+            return decorator  # type: ignore[return-value]
+        if self.exclude_debuggers is not None and pwndbg.dbg.name() in self.exclude_debuggers:
+
+            def decorator(*args, **kwargs):
+                raise InvalidDebuggerError(
+                    f"This command cannot be used in {pwndbg.dbg.name()}.\n"
+                    f"It is invalid for {self.exclude_debuggers}."
+                )
+
+            return decorator  # type: ignore[return-value]
+
+        # Create aliases as seperate CommandObj's and register
+        # them seperately.
+        for alias in self.aliases:
+            CommandObj(
+                function,
+                self.parser,
+                alias,
+                self.category,
+                True,
+                [],
+            )
+        # Since CommandObj has __call__ defined, an instance of it is a
+        # callable object (which essentially decorates the function).
+        return CommandObj(
+            function,
+            self.parser,
+            self.command_name,
+            self.category,
+            False,
+            self.aliases,
```

Comment:
Random thought: since we pass aliases there anyway, wouldn't it be better for the `CommandObj` call `self.register_command(command_name)` for command name + each of aliases instead of us re-creating the whole object for each alias?